pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -413,21 +416,63 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
+    ZNRecord znRecord = _pinotHelixResourceManager.getZnRecord(offlineTableName, segmentName);
 
-    if (offlineSegmentZKMetadata == null) {
+    if (znRecord == null) {
       // Creating new segment, not refreshing
       downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
       _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
       response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
       return response;
     }
+
+    OfflineSegmentZKMetadata offlineSegmentZKMetadata = new OfflineSegmentZKMetadata(znRecord);
+
+    long minionExpectedCrc = -1L;
+    boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
+    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    if (isMinionConfigured) {
+      // Minion configured use case so that we write to zk sparingly
+      String segmentPushStatus = offlineSegmentZKMetadata.getSegmentPushStatus();
+      SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
+      Date dt = new Date();
+      String currentHumanReadableTime = sdf.format(dt);
+      offlineSegmentZKMetadata.setSegmentPushStatus("COMMITTING-" + currentHumanReadableTime);
+
+      if (segmentPushStatus != null) {
+        if (tooLate(currentHumanReadableTime, sdf)) {
+          LOGGER.warn("Segment was not properly committed, replacing segment {} for table {}", segmentName, tableName);
+        } else {
+          throw new ControllerApplicationException(LOGGER, "Please retry later, segment upload in progress", Response.Status.INTERNAL_SERVER_ERROR);
+        }
+      }
+
+      if (!_pinotHelixResourceManager.updateZkMetadataAtomically(offlineSegmentZKMetadata, znRecord.getVersion())) {
+        throw new ControllerApplicationException(LOGGER, "Retry later. ZK update failed.", Response.Status.INTERNAL_SERVER_ERROR);
+      }
+
+      try {
+        minionExpectedCrc = Long.parseLong(headers.getHeaderString(HttpHeaders.IF_MATCH));
+      } catch (NumberFormatException e) {
+        LOGGER.warn("Could not parse crc {}", minionExpectedCrc, e);
+      }
+    }
+
     if (downloadUrl == null) {
       downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
     }
-    // TODO: this will read table configuration again from ZK. We should optimize that
-    response = _pinotHelixResourceManager.refreshSegment(segmentMetadata, downloadUrl);
+
+    if (ifRefreshAnExistedSegment(
+        segmentMetadata, offlineSegmentZKMetadata, offlineTableName, segmentName, minionExpectedCrc, minionHeaderPresent)) {
+      response = _pinotHelixResourceManager.refreshSegment(segmentMetadata, downloadUrl);
+    } else {
+      String msg =
+          "Not refreshing identical segment " + segmentName + "of table " + offlineTableName + " with creation time " + segmentMetadata.getIndexCreationTime() + " and crc "
```

Comment:
we don't know that it is identical. The method ifRefreshAnExistedSegment() also refuses segments that are older, right? (And rename that method to 'shouldRefreshSegment()' please, and move the log to that method, so that it logs the right reason why a segment was not refreshed.)